Catch rebase aborted errors

This error should not fail the build, but it should be logged and continue.

```
##[error]Error: Error while rebasing for {"ownerName":"Label305","repoName":"MyRepo","number":425,"draft":false,"rebaseable":true,"mergeableState":"behind","labels":["autorebase:opt-in"]}: Error: Rebase aborted because the head branch changed.
```

--- a/src/Rebaser/__tests__/rebaser.test.ts
+++ b/src/Rebaser/__tests__/rebaser.test.ts
@@ -1,0 +1,109 @@
+import {Rebaser} from '../rebaser';
+import {GithubRebase} from '../githubRebase';
+
+class TestGithubRebase implements GithubRebase {
+    result: Promise<string> = Promise.reject();
+
+    rebasePullRequest(owner: string, pullRequestNumber: number, repo: string): Promise<string> {
+        return this.result;
+    }
+}
+
+const testGithubRebase = new TestGithubRebase();
+const rebaser = new Rebaser(testGithubRebase);
+
+test('An empty list of pull requests completes successfully', async () => {
+    /* Given */
+    let error = null;
+
+    /* When */
+    try {
+        await rebaser.rebasePullRequests([]);
+    } catch (e) {
+        error = e;
+    }
+
+    /* Then */
+    expect(error).toBeNull();
+});
+
+test('A single succeeding rebase completes successfully', async () => {
+    /* Given */
+    testGithubRebase.result = Promise.resolve('success');
+
+    let error = null;
+
+    /* When */
+    try {
+        await rebaser.rebasePullRequests([
+            {
+                ownerName: 'owner',
+                repoName: 'repo',
+                number: 3,
+                draft: false,
+                rebaseable: true,
+                mergeableState: 'behind',
+                labels: [],
+            },
+        ]);
+    } catch (e) {
+        error = e;
+    }
+
+    /* Then */
+    expect(error).toBeNull();
+});
+
+test('Failing rebase due to head base change completes successfully', async () => {
+    /* Given */
+    testGithubRebase.result = Promise.reject('Rebase aborted because the head branch changed');
+
+    let error = null;
+
+    /* When */
+    try {
+        await rebaser.rebasePullRequests([
+            {
+                ownerName: 'owner',
+                repoName: 'repo',
+                number: 3,
+                draft: false,
+                rebaseable: true,
+                mergeableState: 'behind',
+                labels: [],
+            },
+        ]);
+    } catch (e) {
+        error = e;
+    }
+
+    /* Then */
+    expect(error).toBeNull();
+});
+
+test('Failing rebase due unknown failure errors', async () => {
+    /* Given */
+    testGithubRebase.result = Promise.reject('Some unknown error');
+
+    let error = null;
+
+    /* When */
+    try {
+        await rebaser.rebasePullRequests([
+            {
+                ownerName: 'owner',
+                repoName: 'repo',
+                number: 3,
+                draft: false,
+                rebaseable: true,
+                mergeableState: 'behind',
+                labels: [],
+            },
+        ]);
+    } catch (e) {
+        error = e;
+    }
+
+    /* Then */
+    expect(error).not.toBeNull();
+});

--- a/src/Rebaser/githubRebase.ts
+++ b/src/Rebaser/githubRebase.ts
@@ -1,0 +1,23 @@
+import {rebasePullRequest} from 'github-rebase/lib';
+import {Octokit} from '@octokit/rest';
+
+export interface GithubRebase {
+    rebasePullRequest(owner: string, pullRequestNumber: number, repo: string): Promise<string>;
+}
+
+export class RealGithubRebase implements GithubRebase {
+    private readonly octokit: Octokit;
+
+    constructor(octokit: Octokit) {
+        this.octokit = octokit;
+    }
+
+    public async rebasePullRequest(owner: string, pullRequestNumber: number, repo: string): Promise<string> {
+        return rebasePullRequest({
+            octokit: this.octokit,
+            owner: owner,
+            pullRequestNumber: pullRequestNumber,
+            repo: repo,
+        });
+    }
+}

--- a/src/Rebaser/rebaser.ts
+++ b/src/Rebaser/rebaser.ts
@@ -1,18 +1,16 @@
-import {PullRequestInfo} from './pullrequestinfo';
-import {info} from '@actions/core';
-import {rebasePullRequest} from 'github-rebase/lib';
-import {Octokit} from '@octokit/rest';
-import {GitHub} from '@actions/github';
+import {PullRequestInfo} from '../pullrequestinfo';
+import {info, warning} from '@actions/core';
+import {GithubRebase} from './githubRebase';
 
 /**
  * Uses [github-rebase](https://github.com/tibdex/github-rebase)
  * to rebase pull requests.
  */
 export class Rebaser {
-    private github: GitHub;
+    private githubRebase: GithubRebase;
 
-    constructor(github: GitHub) {
-        this.github = github;
+    constructor(githubRebase: GithubRebase) {
+        this.githubRebase = githubRebase;
     }
 
     public async rebasePullRequests(pullRequests: PullRequestInfo[]): Promise<void> {
@@ -24,15 +22,14 @@ export class Rebaser {
     private async rebase(pullRequest: PullRequestInfo) {
         info(`Rebasing pull request ${JSON.stringify(pullRequest)}`);
         try {
-            await rebasePullRequest({
-                octokit: (this.github as unknown) as Octokit,
-                owner: pullRequest.ownerName,
-                pullRequestNumber: pullRequest.number,
-                repo: pullRequest.repoName,
-            });
+            await this.githubRebase.rebasePullRequest(pullRequest.ownerName, pullRequest.number, pullRequest.repoName);
 
             info(`${JSON.stringify(pullRequest)} was successfully rebased.`);
         } catch (e) {
+            if (String(e).includes('Rebase aborted because the head branch changed')) {
+                warning(`Rebase aborted because the head branch changed for ${JSON.stringify(pullRequest)}`);
+                return;
+            }
             throw new Error(`Error while rebasing for ${JSON.stringify(pullRequest)}: ${String(e)}`);
         }
     }

--- a/src/index.ts
+++ b/src/index.ts
@@ -2,7 +2,7 @@ import {getInput, setFailed} from '@actions/core';
 import {context, GitHub} from '@actions/github';
 import Webhooks from '@octokit/webhooks';
 import {EligiblePullRequestsRetriever} from './EligiblePullRequests/eligiblePullRequestsRetriever';
-import {Rebaser} from './rebaser';
+import {Rebaser} from './Rebaser/rebaser';
 import {TestableEligiblePullRequestsRetriever} from './EligiblePullRequests/testableEligiblePullRequestsRetriever';
 import {GithubPullRequestInfoProvider} from './Github/githubPullRequestInfoProvider';
 import {GithubGetPullRequestService} from './Github/Api/getPullRequestService';
@@ -10,6 +10,8 @@ import {GithubListPullRequestsService} from './Github/Api/listPullRequestsServic
 import {GithubLabelPullRequestService} from './Github/githubLabelPullRequestService';
 import {GithubOpenPullRequestsProvider} from './Github/githubOpenPullRequestsProvider';
 import {Labeler} from './NonRebaseablePullRequests/labeler';
+import {Octokit} from '@octokit/rest';
+import {RealGithubRebase} from './Rebaser/githubRebase';
 
 async function run(): Promise<void> {
     try {
@@ -21,7 +23,7 @@ async function run(): Promise<void> {
         const eligiblePullRequestsRetriever: EligiblePullRequestsRetriever = new TestableEligiblePullRequestsRetriever(
             openPullRequestsProvider,
         );
-        const rebaser = new Rebaser(github);
+        const rebaser = new Rebaser(new RealGithubRebase((github as unknown) as Octokit));
         const labeler = new Labeler(openPullRequestsProvider, new GithubLabelPullRequestService(github));
 
         const payload = context.payload as Webhooks.WebhookPayloadPush;

